logstash-core/lib/logstash/agent.rb

```diff
@@ -310,34 +319,35 @@ def converge_state(pipeline_actions)
         # that we currently have.
         begin
           logger.debug("Executing action", :action => action)
-          action_result = action.execute(self, pipelines)
+          action_result = action.execute(self, @pipelines_registry)
```

Comment:
Not introduced by your PR, but I find the behaviour of passing the entire `self` to a method along with an instance variable of same _self_ to be a bit fishy (previously this had one more small step of indirection, passing the ivar by means of an attribute accessor). It looks like these are largely done to enable the `PipelineAction` to new-up a pipeline that links back to the agent.

No action requested on this, just something to think about.

Reply:
totally agreed. IMO the problem here is passing the who agent object into the action which is now used for 2 purposes I believe: 1- do the `agent.exclusive` on pipeline initialization, which we could get rid with your suggestion above and 2- the use of `ExecutionContextFactory` in the pipeline. 
I think we could refactor both to get rid of the requirement of passing the whole agent. I'd like to defer that to another PR.